draft-hendrickson-privacypass-public-metadata-issuance.md

```diff
@@ -28,25 +28,247 @@ author:
+
+The public metadata issuance protocol differs from the protocol in
+{{Section 6 of !PROTOCOL}} in that the issuance and redemption protocols carry metadata provided by the Client and visible to the Attester, Issuer, and Origin. This means Clients can set arbitrary metadata when requesting a token, but specific values of metadata may be rejected by any of Attester, Issuer, or Origin. Similar to a token nonce, metadata is cryptographically bound to a token and cannot be altered.
+
+Clients provide the following as input to the issuance protocol:
+
+- Issuer Request URI: A URI to which token request messages are sent. This can
+  be a URL derived from the "issuer-request-uri" value in the Issuer's
+  directory resource, or it can be another Client-configured URL. The value
+  of this parameter depends on the Client configuration and deployment model.
+  For example, in the 'Split Origin, Attester, Issuer' deployment model, the
+  Issuer Request URI might be correspond to the Client's configured Attester,
+  and the Attester is configured to relay requests to the Issuer.
+- Issuer name: An identifier for the Issuer. This is typically a host name that
+  can be used to construct HTTP requests to the Issuer.
+- Issuer Public Key: `pkI`, with a key identifier `token_key_id` computed as
+  described in {{public-issuer-configuration}}.
+- Challenge value: `challenge`, an opaque byte string. For example, this might
+  be provided by the redemption protocol in {{AUTHSCHEME}}.
+- Public Metadata: `metadata`, an opaque byte string of length at most 2<sup>16-1</sup> bytes.
+
+Given this configuration and these inputs, the two messages exchanged in
+this protocol are described below. The constant `Nk` is defined as 256 for token type 0xDA7A.
+
+## Client-to-Issuer Request {#public-request}
+
+The Client first creates an issuance request message for a random value
+`nonce` using the input challenge and Issuer key identifier as follows:
+
+~~~
+nonce = random(32)
+challenge_digest = SHA256(challenge)
+token_input = concat(0xDA7A, // Token type field is 2 bytes long
+                     len(metadata),
+                     metadata,
+                     nonce,
+                     challenge_digest,
+                     token_key_id)
+blinded_msg, blind_inv = Blind(pkI, PrepareIdentity(token_input), metadata)
+~~~
+
+Where  `PrepareIdentity` is defined in {{Section 6 of !PBLINDRSA}} and `Blind` is defined in {{Section 4.2 of !PBLINDRSA}}
+
+The Client stores the nonce, challenge_digest, and metadata values locally for use
+when finalizing the issuance protocol to produce a token (as described
+in {{public-finalize}}).
+
+The Client then creates a TokenRequest structured as follows:
+
+~~~
+struct {
+  uint16_t token_type = 0xDA7A; /* Type Public Metadata Blind RSA (2048-bit) */
+  uint8_t truncated_token_key_id;
+  opaque metadata<1..2^16-1>;
```

Comment:
@smhendrickson I suggested the length be two bytes because that's how it's encoded here.

Reply:
Yup I noted this. I was initially thinking this was arbitrary length, so the len output should be as well. 2-byte len makes sense to me, and I accepted the 4->2 bug fix